### ISSUE TITLE
Custom coala bear: Use Closes vs Fixes ddepending on whether issue has the "bug" label

### ISSUE THREAD

**Issue by Makman2** [coala/coala#5155]
Because this is very specific to coala, I've filed the issue here. Maybe this should be in a completely new repo, like `custom-bears` that are only relevant for our org.

What the bear has to do? It checks the commit message whether it contains an issue reference, and makes a HTTP request on the referenced issue. If it's marked as `bug`, `Fixes` has to be used as a keyword, otherwise `Closes`. Should also work on multiple issue references.

Maybe we can derive the bear somehow from `GitCommitBear`? Or reuse it somehow?

**Comment by aashraybhandar1** [comment]
I'm interested in this issue @Makman2 .. Should this be a new bear or part of GitCommitBear?

**Comment by Makman2** [comment]
I'm not sure, please do some considerations.

**Comment by virresh** [comment]
@Makman2
I am interested in this one. 
I have the following idea for this:
We can indeed extend the GitCommitBear and extract issue number and keyword from the issue link.
We can use the GitHub Api v3 to get labels from the issue number, and we can then process them accordingly. (like this one : https://api.github.com/repos/coala/coala/issues/5155/labels )

The only point I feel will be a little messy would be in case we do not provide complete links, for in that case it would be difficult to determine which is the repository to which the issue belongs, and in that case we might need to hardcode a few things, which is probably bad.

Although since we demand a complete issue reference be provided in the commit message, I don't think the above will be an issue.

**Comment by Makman2** [comment]
> We can indeed extend the GitCommitBear

The only problem is that this functionality might be too coala-related, and thus not belonging into our generic `GitCommitBear`.

> The only point I feel will be a little messy would be in case we do not provide complete links, for in that case it would be difficult to determine which is the repository to which the issue belongs, and in that case we might need to hardcode a few things, which is probably bad.

In our coala projects we require full links in our coafiles. However, it could be maybe a separate issue to implement a setting for the hoster, like `hoster="GitHub"` or `hoster="GitLab"`, that completes all needed info. That should integrate quite well into the current `GitCommitBear` code.

**Comment by virresh** [comment]
Yeah, so what I realised while making an HgCommitBear is that both Hg and Git bears resembled each other closely. At jayvdb's suggestion, I realised we can actually make a new _CommitBear and inherit GitCommitBear and HgCommitBear to avoid code duplication. 
On the same lines, we can again inherit from the _CommitBear and make an IssueReferenceBear which would do the required functionality.
Like we would provide this bear the issue labels to take care (probably using some flag which receives dictionary of labels and corresponding keyword of from .coafile) and the bear will classify accordingly as to what kind of reference should be used for the supported hosts. 

Or maybe we can simply create another global bear IssueReferenceBear, which will do all the computations, since probably IssueReferenceBear and _CommitBear won't have much in common.

Also for the hoster one, probably we need to have a hoster from one of the supported hosts, since every hosting service will have it's own api for getting the issue labels. (Indeed very similar to how we treat the keywords Fix/Close etc).

**Comment by Makman2** [comment]
Okay inheritance sounds like a good option in case our existing bears have much in common. :+1.

**Comment by jayvdb** [comment]
IMO this would be easiest done as a custom gitmate plugin, as it already has all of this information handy.  Hence https://github.com/coala/projects/issues/488

However an aspect which is very general is that currently the bear doesnt even check the issue exists or is open.  Hence https://github.com/coala/coala-bears/issues/2369

Also this specific problem of Fixes/Closes and a label on the issue, that could be made generic by building the rule that a configured regexes for the commit message must have a configured label on the issue. e.g.
```
commit_issue_label_patters = 
  /Fixes http/ : bug
```

**Comment by RaiVaibhav** [comment]
@Makman2  @jayvdb @virresh  If I understand correctly than why can't we directly [use IGitt (_get_keyword_issue)](https://gitlab.com/gitmate/open-source/IGitt/blob/master/IGitt/Interfaces/MergeRequest.py#L178)
and accordingly go to the the issue url and find their labels directly ( and of course we can check whether `issue` is a `PR` or not ), I don't understand whether you want this functionality to work it as a Bear so that everyone can use this directly or something else?

**Comment by RaiVaibhav** [comment]
And somehow if we don't have any feature in IGitt then IMO adding that feature in IGitt will be my first priority so that I can use with a simple one line of code :smiley:

**Comment by Makman2** [comment]
> Anyway, this is not an issue that can be fixed in the coala repo.

Actually I intended writing a custom bear here in this repo. Even if we can't use IGitt directly, we can use HTTP-requests in the worst case scenario.

Imo this issue is valid and still not done yet.
